core/jobs.py

```diff
@@ -165,35 +163,52 @@ def register_completion(cls, job_id, output_list):
 
         model.status_code = STATUS_CODE_COMPLETED
         model.time_finished_msec = utils.get_current_time_in_millisecs()
+        model.output = cls._compress_output_list(output_list)
+        model.put()
+
+        cls._post_completed_hook(job_id)
 
-        # TODO(bhenning): Add tests for this.
-        output_str_list = ['%s' % output_value for output_value in output_list]
+    @classmethod
+    def _compress_output_list(
+            cls, output_list, test_only_max_output_chars=None):
+        """Returns compressed list of strings with a max length of chars.
+
+        Args:
+            output_list: list(*). Collection of objects to be stringified.
+            test_only_max_output_chars: int or None. Used to help test this
+                method.
 
-        # De-duplicate the lines of output since it's not very useful to repeat
-        # them.
-        counter = collections.Counter(list(output_str_list))
-        output_str_frequency_list = [
-            (output_str, counter[output_str]) for output_str in counter]
+        Returns:
+            list(str). The compressed stringified output values.
+        """
+        _MAX_OUTPUT_CHARS = 900000
+        class _OrderedCounter(collections.Counter, collections.OrderedDict):
```

Comment:
I'm not sure I understand the need for this, and am slightly reluctant to introduce it. Why do we care about the order?

If it's just for tests, better to just call `sorted` on the actual/expected outputs instead.